crates/prettier/src/prettier.rs

```diff
@@ -305,14 +305,13 @@ impl Prettier {
                             })
                             .collect();
 
-                        let mut prettier_parser = prettier_settings.parser.as_deref();
-                        if buffer_path.is_none() {
-                            prettier_parser = prettier_parser.or_else(|| buffer_language.and_then(|language| language.prettier_parser_name()));
-                            if prettier_parser.is_none() {
-                                log::error!("Formatting unsaved file with prettier failed. No prettier parser configured for language {buffer_language:?}");
-                                return Err(anyhow!("Cannot determine prettier parser for unsaved file"));
-                            }
-
+                        let mut prettier_parser = buffer_language.and_then(|language| language.prettier_parser_name());
+                        if prettier_parser.is_none() {
+                            prettier_parser = prettier_settings.parser.as_deref();
```

Comment:
Also, this seems not correct: if there's something specifically stated as a parser in settings.json, we have to take that, as a user override.

Am I missing something?

Reply:
The language setting should be higher than the prettier setting, because the user intent to use that syntax to edit. Ideally the syntax and prettier setting should be synced.

Reply:
Great, seems like more tests for you to write then, to demonstrate this.

Reply:
@SomeoneToIgnore can you guide me how to set the language of the buffer in the test? I'm the beginner of Rust 😅

Reply:
It's not about Rust but barely the API usage:
* init a language and add it to the registry similar to how the test linked above
* https://github.com/zed-industries/zed/blob/f9990b42fa3ef81b53d0b0f42c88836e319758f1/crates/language/src/buffer.rs#L962

Reply:
```rust
#[gpui::test]
async fn test_prettier_formats_with_language_specified_parser(cx: &mut gpui::TestAppContext) {
    init_test(cx, |settings| {
        settings.defaults.formatter = Some(language_settings::SelectedFormatter::List(
            FormatterList(vec![Formatter::Prettier].into()),
        ))
    });

    let fs = FakeFs::new(cx.executor());
    fs.insert_file("/file.custom", Default::default()).await;

    let project = Project::test(fs, ["/file.custom".as_ref()], cx).await;
    let language_registry = project.read_with(cx, |project, _| project.languages().clone());

    let language = Arc::new(Language::new(
        LanguageConfig {
            name: "TypeScript".into(),
            matcher: LanguageMatcher {
                path_suffixes: vec!["ts".to_string()],
                ..Default::default()
            },
            ..Default::default()
        },
        Some(tree_sitter_rust::LANGUAGE.into()),
    ));

    language_registry.add(language.clone());

    update_test_language_settings(cx, |settings| {
        settings.defaults.prettier = Some(PrettierSettings {
            allowed: true,
            ..PrettierSettings::default()
        });
    });

    let test_plugin = "test_plugin";
    let _ = language_registry.register_fake_lsp(
        "TypeScript",
        FakeLspAdapter {
            prettier_plugins: vec![test_plugin],
            ..Default::default()
        },
    );

    let prettier_format_suffix = project::TEST_PRETTIER_FORMAT_SUFFIX;
    let buffer = project
        .update(cx, |project, cx| {
            project.open_local_buffer("/file.custom", cx)
        })
        .await
        .unwrap();

    buffer.update(cx, |buffer, _cx| {
        buffer.set_language(Some(language), _cx);
    });

    let buffer_text = "one\ntwo\nthree\n";
    let buffer = cx.new_model(|cx| MultiBuffer::singleton(buffer, cx));

    let (editor, cx) = cx.add_window_view(|cx| build_editor(buffer, cx));
    editor.update(cx, |editor, cx| editor.set_text(buffer_text, cx));

    editor
        .update(cx, |editor, cx| {
            editor.perform_format(
                project.clone(),
                FormatTrigger::Manual,
                FormatTarget::Buffer,
                cx,
            )
        })
        .unwrap()
        .await;
    assert_eq!(
        editor.update(cx, |editor, cx| editor.text(cx)),
        buffer_text.to_string() + prettier_format_suffix,
        "Test prettier formatting was not applied to the original buffer text",
    );

    update_test_language_settings(cx, |settings| {
        settings.defaults.formatter = Some(language_settings::SelectedFormatter::Auto)
    });
    let format = editor.update(cx, |editor, cx| {
        editor.perform_format(
            project.clone(),
            FormatTrigger::Manual,
            FormatTarget::Buffer,
            cx,
        )
    });
    format.await.unwrap();
    assert_eq!(
        editor.update(cx, |editor, cx| editor.text(cx)),
        buffer_text.to_string() + prettier_format_suffix + "\n" + prettier_format_suffix,
        "Autoformatting (via test prettier) was not applied to the original buffer text",
    );
}
```

```
running 1 test
thread 'editor_tests::test_prettier_formats_custom_extension_with_language_specified_parser' panicked at crates/editor/src/editor_tests.rs:10741:5:
assertion `left == right` failed: Test prettier formatting was not applied to the original buffer text
  left: "one\ntwo\nthree\n"
 right: "one\ntwo\nthree\n\nformatted by test prettier"
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
test editor_tests::test_prettier_formats_custom_extension_with_language_specified_parser ... FAILED
```

Thanks for advice, I did trace the code but still can't figure out why the prettier not apply correctly. It looks like Buffer and MultiBuffer not synced. Or the timing what I set the language to buffer is wrong.

Reply:
looks like test have different implementation

// prettier.rs
```rust
            Self::Test(_) => Ok(buffer
                .update(cx, |buffer, cx| {
                    match buffer
                        .language()
                        .map(|language| language.lsp_id())
                        .as_deref()
                    {
                        Some("rust") => anyhow::bail!("prettier does not support Rust"),
                        Some(_other) => {
                            let formatted_text = buffer.text() + FORMAT_SUFFIX;
                            Ok(buffer.diff(formatted_text, cx))
                        }
                        None => panic!("Should not format buffer without a language with prettier"),
                    }
                })??
                .await),
```

Reply:
It's a good idea to run the problematic tests with logs enabled, to see what it prints, on Unix: `env RUST_LOG=info ...rest of the test launching command`

The new tests prints this:
```
[2024-11-18T12:42:41Z ERROR project] Default prettier is not installed and cannot be started
```

Now we have to understand why does this happen.
You can take the debugger, place the breakpoint at the place of logging and walk around, or place `dbg!("foo", &variable);` somewhere in the code and see what it prints on test reruns.

I've added this:
```
diff --git a/crates/editor/src/editor_tests.rs b/crates/editor/src/editor_tests.rs
index 4469b2e614..bf55194699 100644
--- a/crates/editor/src/editor_tests.rs
+++ b/crates/editor/src/editor_tests.rs
@@ -10576,6 +10576,105 @@ async fn test_completions_in_languages_with_extra_word_characters(cx: &mut gpui:
     });
 }
 
+#[gpui::test]
+async fn test_prettier_formats_with_language_specified_parser(cx: &mut gpui::TestAppContext) {
+    init_test(cx, |settings| {
+        settings.defaults.formatter = Some(language_settings::SelectedFormatter::List(
+            FormatterList(vec![Formatter::Prettier].into()),
+        ))
+    });
+
+    let fs = FakeFs::new(cx.executor());
+    fs.insert_file("/file.custom", Default::default()).await;
+
+    let project = Project::test(fs, ["/file.custom".as_ref()], cx).await;
+    let language_registry = project.read_with(cx, |project, _| project.languages().clone());
+
+    let language = Arc::new(Language::new(
+        LanguageConfig {
+            name: "TypeScript".into(),
+            matcher: LanguageMatcher {
+                path_suffixes: vec!["ts".to_string()],
+                ..Default::default()
+            },
+            ..Default::default()
+        },
+        Some(tree_sitter_rust::LANGUAGE.into()),
+    ));
+
+    language_registry.add(language.clone());
+
+    update_test_language_settings(cx, |settings| {
+        settings.defaults.prettier = Some(PrettierSettings {
+            allowed: true,
+            ..PrettierSettings::default()
+        });
+    });
+
+    let test_plugin = "test_plugin";
+    let _ = language_registry.register_fake_lsp(
+        "TypeScript",
+        FakeLspAdapter {
+            prettier_plugins: vec![test_plugin],
+            ..Default::default()
+        },
+    );
+
+    let prettier_format_suffix = project::TEST_PRETTIER_FORMAT_SUFFIX;
+    let buffer = project
+        .update(cx, |project, cx| {
+            project.open_local_buffer("/file.custom", cx)
+        })
+        .await
+        .unwrap();
+
+    buffer.update(cx, |buffer, _cx| {
+        buffer.set_language(Some(language), _cx);
+    });
+
+    let buffer_text = "one\ntwo\nthree\n";
+    let buffer = cx.new_model(|cx| MultiBuffer::singleton(buffer, cx));
+
+    let (editor, cx) = cx.add_window_view(|cx| build_editor(buffer, cx));
+    editor.update(cx, |editor, cx| editor.set_text(buffer_text, cx));
+
+    dbg!("?????????????????");
+    editor
+        .update(cx, |editor, cx| {
+            editor.perform_format(
+                project.clone(),
+                FormatTrigger::Manual,
+                FormatTarget::Buffer,
+                cx,
+            )
+        })
+        .unwrap()
+        .await;
+    assert_eq!(
+        editor.update(cx, |editor, cx| editor.text(cx)),
+        buffer_text.to_string() + prettier_format_suffix,
+        "Test prettier formatting was not applied to the original buffer text",
+    );
+
+    update_test_language_settings(cx, |settings| {
+        settings.defaults.formatter = Some(language_settings::SelectedFormatter::Auto)
+    });
+    let format = editor.update(cx, |editor, cx| {
+        editor.perform_format(
+            project.clone(),
+            FormatTrigger::Manual,
+            FormatTarget::Buffer,
+            cx,
+        )
+    });
+    format.await.unwrap();
+    assert_eq!(
+        editor.update(cx, |editor, cx| editor.text(cx)),
+        buffer_text.to_string() + prettier_format_suffix + "\n" + prettier_format_suffix,
+        "Autoformatting (via test prettier) was not applied to the original buffer text",
+    );
+}
+
 #[gpui::test]
 async fn test_document_format_with_prettier(cx: &mut gpui::TestAppContext) {
     init_test(cx, |settings| {
diff --git a/crates/project/src/lsp_store.rs b/crates/project/src/lsp_store.rs
index 0723ba689b..fab42cd09e 100644
--- a/crates/project/src/lsp_store.rs
+++ b/crates/project/src/lsp_store.rs
@@ -1251,6 +1251,7 @@ impl LspStore {
         new_language: Arc<Language>,
         cx: &mut ModelContext<Self>,
     ) {
+        dbg!("1");
         buffer.update(cx, |buffer, cx| {
             if buffer.language().map_or(true, |old_language| {
                 !Arc::ptr_eq(old_language, &new_language)
@@ -1259,6 +1260,7 @@ impl LspStore {
             }
         });
 
+        dbg!("2");
         let buffer_file = buffer.read(cx).file().cloned();
         let settings =
             language_settings(Some(new_language.name()), buffer_file.as_ref(), cx).into_owned();
@@ -1273,7 +1275,7 @@ impl LspStore {
             None
         };
 
-        if settings.prettier.allowed {
+        if dbg!(settings.prettier.allowed) {
             if let Some(prettier_plugins) = prettier_store::prettier_plugins_for_language(&settings)
             {
                 let prettier_store = self.as_local().map(|s| s.prettier_store.clone());
diff --git a/crates/project/src/prettier_store.rs b/crates/project/src/prettier_store.rs
index c7ac0ffd0b..be4112785e 100644
--- a/crates/project/src/prettier_store.rs
+++ b/crates/project/src/prettier_store.rs
@@ -456,6 +456,7 @@ impl PrettierStore {
         plugins: impl Iterator<Item = Arc<str>>,
         cx: &mut ModelContext<Self>,
     ) {
+        dbg!("???????@@@@@@@@@@@@@??????");
         if cfg!(any(test, feature = "test-support")) {
             self.default_prettier.installed_plugins.extend(plugins);
             self.default_prettier.prettier = PrettierInstallation::Installed(PrettierInstance {

```


And see that prettier test installation had never happened for this language.
Feels that for this particular test, we have to take a project and call

https://github.com/zed-industries/zed/blob/9260abafbaa3eddb3922da3555d6dbe9903b05d0/crates/project/src/project.rs#L2399

instead of our language settings that I've recommended.

```
project.update(cx, |project, cx| {
    project.set_language_for_buffer(&buffer, language, cx);
});
```

Which seem to make the test pass.

Reply:
Thank you! I will update it in the coming days.